x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/RemoteIndicesPermission.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.authz.permission;
+
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xpack.core.security.authz.RestrictedIndices;
+import org.elasticsearch.xpack.core.security.authz.privilege.IndexPrivilege;
+import org.elasticsearch.xpack.core.security.support.Automatons;
+import org.elasticsearch.xpack.core.security.support.StringMatcher;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public record RemoteIndicesPermission(List<RemoteIndicesGroup> remoteIndicesGroups) {
+
+    public static final RemoteIndicesPermission NONE = new RemoteIndicesPermission(List.of());
+
+    public RemoteIndicesPermission forCluster(final String remoteClusterAlias) {
+        return new RemoteIndicesPermission(
+            remoteIndicesGroups.stream()
+                .filter(remoteIndicesGroup -> remoteIndicesGroup.checkRemoteClusterAlias(remoteClusterAlias))
+                .toList()
+        );
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+        final Map<Set<String>, List<IndicesPermission.Group>> remoteIndicesGroups;
+
+        public Builder() {
+            this.remoteIndicesGroups = new HashMap<>();
+        }
+
+        public Builder addGroup(
+            final Set<String> remoteClusterAliases,
+            final IndexPrivilege privilege,
+            final FieldPermissions fieldPermissions,
+            final @Nullable Set<BytesReference> query,
+            final boolean allowRestrictedIndices,
+            final String... indices
+        ) {
+            remoteIndicesGroups.computeIfAbsent(remoteClusterAliases, k -> new ArrayList<>())
+                .add(
+                    new IndicesPermission.Group(
+                        privilege,
+                        fieldPermissions,
+                        query,
+                        allowRestrictedIndices,
+                        // Deliberately passing EMPTY here since *which* indices are restricted is determined not on the querying cluster
+                        // but rather on the fulfilling cluster
+                        new RestrictedIndices(Automatons.EMPTY),
```

Comment:
Agreed that we will need to revisit this. I think when we get to implementing the pre-flight authorization check on the querying cluster it will become clearer what to do about restricted indices here. Will keep your concern in mind. Thanks for raising!